lib/seed/runner.rb

```diff
@@ -0,0 +1,144 @@
+    attr_reader :logger
+    attr_reader :start_time
+    attr_accessor :counts
+    attr_accessor :total_counts
+
+    def self.call(*args)
+      new(*args).call
+    end
+
+    delegate :scale_factor, to: :config
+    delegate :distance_of_time_in_words, to: Seed::Helpers
+
+    def initialize(config: Seed::Config.new)
+      @counts = {}
+      @total_counts = {}
+      @config = config
+      @logger = Rails.logger.child(class: self.class.name)
+      @start_time = Time.current
+      puts "Starting #{self.class} with #{config.type} configuration"
+    end
+
+    def create_progress_bar
+      ProgressBar.create(
+        format: "%t |%E | %b\u{15E7}%i %p%% | %a",
+        remainder_mark: "\u{FF65}",
+        title: "Seeding Facilities",
+        total: Facility.count
+      )
+    end
+
+    def call
+      result = FacilitySeeder.call(config: config)
+      total_counts[:facility] = result&.ids&.size || 0
+      UserSeeder.call(config: config)
+
+      puts "Starting to seed patient data for #{Facility.count} facilities..."
+
+      progress = create_progress_bar
+
+      results = seed_patients(progress)
+      results.each { |hsh| counts[hsh.delete(:facility)] = hsh }
+      hsh = sum_facility_totals
+      total_counts.merge!(hsh)
+
+      msg = "⭐️  Seed complete! Elasped time #{distance_of_time_in_words(start_time, Time.current, include_seconds: true)} ⭐️"
+      puts msg
+      logger.info msg: msg, counts: counts
+      [counts, total_counts]
+    end
+
+    def seed_patients(progress)
+      parallel_options = {
+        finish: lambda do |item, i, result|
+          slug, facility_size = item[1], item[2]
+          progress.log("Finished facility: [#{slug}, #{facility_size}] counts: #{result}")
+          progress.increment
+        end
+      }
+      parallel_options[:in_processes] = 0 if Rails.env.test?
+
+      facility_info = Facility.pluck(:id, :slug, :facility_size)
+      Parallel.map(facility_info, parallel_options) do |(facility_id, slug, facility_size)|
+        benchmark("Seeding records for facility #{slug}") do
+          result = {facility: slug}
+          facility = Facility.find(facility_id)
+          user = facility.users.find_by!(role: config.seed_generated_active_user_role)
+          # Set a "birth date" for the Facility that patient records will be based from
+          facility_birth_date = Faker::Time.between(from: 3.years.ago, to: 1.day.ago)
+          benchmark("[#{slug} Seeding patients for a #{facility_size} facility") do
+            patients = patients_to_create(facility_size).times.map { |num|
+              build_patient(user, oldest_registration: facility_birth_date)
+            }
+            patient_result = Patient.import(patients, recursive: true)
+            result[:patient] = patient_result.ids.size
+          end
+          patient_info = facility.assigned_patients.pluck(:id, :recorded_at)
+          bp_result = BloodPressureSeeder.call(config: config, facility: facility, user: user)
```

Comment:
Does this mean that _all_ the BPs for a patient will be at the same facility, and recorded by the same user?